### PR TITLE
Add option to run syncthing when tethering is enabled

### PR DESCRIPTION
**Pull request by grimpy**
Listen to event wifi ap changed to start syncthing

### CODE REVIEW

File: src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java [hunk 1]
```diff
@@ -13,19 +14,27 @@
...
-        ConnectivityManager cm =
-                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
-        NetworkInfo wifiInfo = cm.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
-        boolean isWifiConnected = (wifiInfo != null && wifiInfo.isConnected());
-        Log.v(TAG, "Received wifi " + (isWifiConnected ? "connected" : "disconnected") + " event");
         Intent i = new Intent(context, SyncthingService.class);
-        i.putExtra(DeviceStateHolder.EXTRA_HAS_WIFI, isWifiConnected);
+        String action = intent.getAction();
+        if (action.equals("android.net.wifi.WIFI_AP_STATE_CHANGED")) {
+            int state = intent.getIntExtra("wifi_state", 0);
```
Comment (Zillode) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java]: 0 = WIFI_AP_STATE_DISABLED?

**Comment by Zillode** [comment]
Cool:)

**Comment by Zillode** [comment]
@st-jenkins build this please

**Comment by Nutomic** [comment]
So the purpose of this is to sync to other mobile devices that are connected to the hotspot? Just curious, how do you avoid transfers from all devices over mobile data?

Also, I'm kinda worried that the settings are getting too full with this. maybe we should move these into a ListPreference (though not necessarily in this PR).

File: src/main/java/com/nutomic/syncthingandroid/syncthing/DeviceStateHolder.java [hunk 2]
```diff
@@ -54,8 +59,13 @@ public boolean isWifiConnected() {
         return mIsWifiConnected;
     }
 
+    public boolean isAPOn() {
```
Comment (Nutomic) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/DeviceStateHolder.java]: This should be `isApOn` for better readability (I like `isApEnabled` more). Same for other occurences.

File: src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java [hunk 3]
```diff
@@ -13,19 +14,27 @@
 public class NetworkReceiver extends BroadcastReceiver {
 
     private static final String TAG = "NetworkReceiver";
+    private static final int WIFI_AP_STATE_ENABLED = 13;
```
Comment (Nutomic) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java]: There should be a constant in the Android API somewhere?

**Comment by Nutomic** [comment]
`NetworkReceiverTest` is failing, can you fix that? And ideally add one or more test cases for this functionality.

File: src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java (hunk 3 above)
Comment (grimpy) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java]: https://github.com/android/platform_frameworks_base/blob/master/wifi/java/android/net/wifi/WifiManager.java#L216

Hidden.

File: src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java (hunk 1 above)
Comment (grimpy) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java]: Something that is not enabled... disabled is 11 see https://github.com/android/platform_frameworks_base/blob/master/wifi/java/android/net/wifi/WifiManager.java#L206

**Comment by Zillode** [comment]
@st-jenkins add to whitelist

File: src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java (hunk 3 above)
Comment (Nutomic) [inline src/main/java/com/nutomic/syncthingandroid/syncthing/NetworkReceiver.java]: Ah that sucks. We should verify that this works accross Android versions. [These](http://stackoverflow.com/a/12531759) [links](http://stackoverflow.com/a/14405339) suggest that the values are different before ICS.

File: src/main/res/xml/app_settings.xml [hunk 4]
```diff
@@ -20,6 +20,12 @@
             android:defaultValue="false" />
 
         <CheckBoxPreference
+            android:key="sync_only_ap"
+            android:title="@string/sync_only_ap"
+            android:defaultValue="false" />
+
+
```
Comment (Nutomic) [inline src/main/res/xml/app_settings.xml]: Only one empty line please :)

**Comment by Nutomic** [comment]
Sorry I didn't notice your update. Next time, could you post a comment so I get a notification?

[This](https://github.com/syncthing/syncthing-android/pull/341#discussion_r27831491) seems to be still relevant. Also, the option is not disabled if "Always run in Background" is unchecked.

`NetworkReceiverTest#testOnReceive()` is failing. You should add a test case for the new code anyway ;)

Additionally, I just noticed that the wording isn't entirely accurate ("sync only on x"), but that's a seperate problem.

**Comment by Nutomic** [comment]
Closing this, feel free to reopen if you wanna work on merging it.

**Pull request closed without merging**